[#91207216] Add coreOS docker nodes

mtekel opened the pull request:
https://www.pivotaltracker.com/story/show/91207216 

**What**

This PR moves our docker nodes from running on Ubuntu 14.0.4 LTS  to CoreOS Stable channel (723.3.0) 

This pull request should be reviewed in the context of:

**Acceptance criteria**
**Given** a script (or set of instructions) for creating a Tsuru cluster
**And** that the infrastructure has already been provisioned
**When** I execute the script (or follow the instructions)
**Then** a docker node running on CoreOS is added to the Tsuru cluster and able to run applications

**How to test**

Check out this branch and deploy a new environment. Check for the number of coreos nodes in the platform of your choice. SSH into one of the coreos nodes and check that they are running the stable version of coreos. 

This PR only covers half of this story - there is a second PR on tsuru-ansible to be merged. 
https://github.com/alphagov/tsuru-ansible/pull/165

 **Who should review**

Anyone but @mtekel or @dhilton

mtekel committed:
gce variables: add coreos image ID

Add image id of most recent stable coreOS build available on GCE to our variables.

mtekel committed:
gce: add coreos-docker-servers

coreos-docker-servers.tf contains definition of coreos servers which we use to host docker containers for tsuru.

mtekel committed:
Add coreos config template

Add cloud-config template for coreOS nodes.

mtekel committed:
globals: adding docker_count variable

docker_count specifies number of coreOS nodes to deploy. These serve as docker hosts for the user apps. There should be at least 3 (to enable etcd quorum).

mtekel committed:
gce: add ETCD_CLUSTER_ID placeholder

This file is used to store ETCD_CLUSTER_ID. It needs to exist before terraform starts its run and populates it with real token.

mtekel committed:
gce: remove old docker servers

Docker servers are replaced by coreOS, we don't need to provision them anymore.

mtekel committed:
aws: add ETCD_CLUSTER_ID placeholder

Create empty ETCD_CLUSTER_ID placeholder file that would get populated by real cluster token. The file needs to exist before terraform runs (otherwise terraform refuses to run).

mtekel committed:
aws: add coreos images into variables

Define coreOS image AMI IDs.

mtekel committed:
aws: add 3rd zone into variables

Add IP ranges and name for the 3rd zone in ec2 eu-west to the aws variables.tf

mtekel committed:
aws: add coreos docker servers

Add coreos-docker-servers.tf file describing configuration of coreOS docker servers.

mtekel committed:
aws: remove old docker servers

Ubuntu docker servers are now replaced by coreOS and we no longer need to create them.

mtekel committed:
gitignore: add ETCD_CLUSTER_ID files

<aws|gce>/ETCD_CLUSTER_ID is used for storing ETCD cluster discovery ID. This will be different for each environment and is essentially a temporary value for the lifetime of the environment. It should not be committed to github.

dcarley commented on coreos-config.yaml.tpl:
```diff
@@ -0,0 +1,34 @@
+        - name: 50-insecure-registry.conf
+          content: |
+            [Service]
+            Environment=DOCKER_OPTS='--insecure-registry="${docker_regisry_internal_ip}/32"'
```
We should use a hostname here instead of an IP address so that we don't regress [#93690844](https://www.pivotaltracker.com/story/show/93690844).

mtekel committed:
globals: added registry port variable

Define docker registry port in a variable. This is needed to create insecure registy parameter for coreOS docker nodes which require it to include port when it is host name based. (You can have IP (even IP subnet) based insecure registry parameters without port, but hostname requires port).

mtekel committed:
coreOS: use hostname based insecure registry

Instead of specifying IP address (or range) specify registry DNS record:port as insecure registry server for coreOS servers acting as docker nodes.

mtekel commented:
Updated insecure registry parameter to use DNS record...

dcarley committed:
Start docker service on CoreOS boot

Otherwise the service is only started when systemd receives the first TCP
connection on the socket. This results in a small delay for the first
connection, which will slow down our Ansible run (which registers nodes) by
a few seconds for each.

dcarley committed:
Fix typo in var name docker_registry_host

s/regisry/registry/

dcarley committed:
aws: Rename AMI variables

To make it clearer that we have two variables for AMIs, one for Ubuntu and
the other for CoreOS.

dcarley committed:
Revert "gitignore: add ETCD_CLUSTER_ID files"

This reverts commit 42c16dccf97af032cb69e81f364ebfa460fc3bfb.

This wasn't having the desired effect. git won't ignore changes to files
that are also tracked. So I'm removing these to avoid confusion.

Because Terraform requires these files to be present on the first run (so we
can't not commit them) we'll just have to be careful about not committing
any changes back, for now:

    Errors:

      * 1 error(s) occurred:

    * 1 error(s) occurred:

    * file: open ETCD_CLUSTER_ID: no such file or directory in:

    ${file("ETCD_CLUSTER_ID")}

dcarley commented:
@mtekel I've pushed 4 commits that we talked about. Can you make sure that you're happy with them?

mtekel commented:
Fine with updates.

dcarley committed:
gce: Change user prefix for CoreOS sshKeys

This was previously creating a new user called `ubuntu` and inserting the
public key into its `~/.ssh/authorized_keys`, in addition to also adding it
to the `core` user. Changing the prefix here means that we don't create any
additional users.

dcarley commented:
Tested fine on GCE and AWS. @mtekel could you check that you're happy with the last commit (changing username) and then I can merge.

mtekel commented:
Fine with that. Having 'ubuntu' user on coreOS was a bit confusing.
